pkg/volume/util/util.go

```diff
@@ -381,7 +381,10 @@ func GetUniqueVolumeNameFromSpec(
 
 // IsPodTerminated checks if pod is terminated
 func IsPodTerminated(pod *v1.Pod, podStatus v1.PodStatus) bool {
-	return podStatus.Phase == v1.PodFailed || podStatus.Phase == v1.PodSucceeded || (pod.DeletionTimestamp != nil && notRunning(podStatus.ContainerStatuses))
+	// TODO: the guarantees provided by kubelet status are not sufficient to guarantee it's safe to ignore a deleted pod,
+	// even if everything is notRunning (kubelet does not guarantee that when pod status is waiting that it isn't trying
+	// to start a container).
```

Comment:
does it mean that even if IsPodTerminated() return true, after that, pod might still try to start a container?

Reply:
talked to @Random-Liu about the logic here and checking the use case of isPodTerminated in this volume util, I think we do want to modify the function here so that if isPodTeminated() return true, there will be no more containers trying to start.

Reply:
> it could satisfy our use case to modify the function here so that if isPodTeminated() return true, there will be no more containers trying to start.

It is impossible today for components other than the Kubelet (who read the pod api) to know when the kubelet will no longer restart containers, because the Kubelet does not communicate that via pod status in a way that components can use it.  So if you need that, Kubelet needs a change, or you need to weaken your guarantees.

Within the Kubelet, all kubelet sync loops should NOT be using pkg/volume/util (and I have changed code in the kubelet to not use it) and instead use the PodStateProvider style interfaces that the pod worker provides, which *can* guarantee the check.